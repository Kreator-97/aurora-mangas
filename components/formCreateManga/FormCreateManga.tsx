import { FC, FormEvent, useState } from 'react'
import { Toaster } from 'react-hot-toast'

import { useForm } from '../../hooks'
import { Manga, Serie } from '../../interfaces'

const CLOUDINARY_DOMAIN = 'https://res.cloudinary.com'

const formValidations = {
  serie: {
    validation: (value:string) => value.trim() !== '',
    message: 'Este campo no puede estar vacio',
  },
  title: {
    validation: (value:string) => value.trim() !== '',
    message: 'Este campo no puede estar vacio',
  },
  imgURL: {
    validation: (value:string) => value.trim() !== '' && value.startsWith(CLOUDINARY_DOMAIN) ,
    message: 'Este campo debe de ser una URL de Cloudinary válida',
  },
  number: {
    validation: (value:string) => ( !isNaN(Number(value)) ) && value !== '',
    message: 'Este campo debe de ser un número válido',
  },
  price: {
    validation: (value:string) => ( !isNaN(Number(value)) ) && value !== '',
    message: 'Este campo debe de ser un número válido',
  },
  month: {
    validation: (value:string) => ( !isNaN(Number(value)) ) && value !== '',
    message: 'Este campo debe de ser un número válido entre 1 y 12',
  },
  year: {
    validation: (value:string) => ( !isNaN(Number(value)) ) && value !== '',
    message: 'Este campo debe de ser un año superior a 2020',
  },
}

interface Props {
  series : Serie[];
  onSubmit: (formValues:FormCreateManga) => Promise<void>;
  manga?: Manga;
  resetOnSubmit?: boolean;
}

export interface FormCreateManga {
  serie: string;
  imgURL: string;
  number : string;
  price : string;
  title : string;
  month : string;
  year : string;
}

export const FormCreateManga: FC<Props> = ({series, onSubmit, manga, resetOnSubmit = true}) => {
  const [showErrors, setShowErrors] = useState(false)
  const date = manga?.published ? manga.published.split('/') : []
  const mangaYear = date[0]
  const mangaMonth = date [1]

  const { imgURL, number, price, title, serie, month, year, errors, onInputChange, onResetForm } = useForm<FormCreateManga>({
    serie   : manga?.serie.id || '',
    imgURL  : manga?.imgURL|| '',
    number  : manga?.number || '',
    price   : String(manga?.price) || '',
    title   : manga?.title || '',
    month   : mangaMonth || '',
    year    : mangaYear || '',
  }, formValidations)

  const onSubmitEvent = async (e:FormEvent) => {
    e.preventDefault()

    // this validate form before send it to submit
    if( Object.values(errors).some( value => value !== null) ) {
      setShowErrors(true)
      return
    }

    const formValid:FormCreateManga = { serie, title, number, imgURL, price, year, month }

    // we execute onSubmit event handle from parent element
    await onSubmit(formValid)

    setShowErrors(false)
    if( resetOnSubmit ) {
      onResetForm()
    }
  } 

  return (
    <form className='mx-auto mb-4 max-w-lg' onSubmit={onSubmitEvent} data-testid="form-create-manga">
      <div className='flex flex-col gap-4 mb-4'>

        <div className='flex flex-col gap-2'>
          <label htmlFor="serie" className='text-lg'>
              Seleccionar serie
          </label>
          <select
            name="serie"
            id="serie"
            className='input bg-white'
            value={ serie }
            onChange={ onInputChange }
          >
            <option
              value=""
              defaultChecked
              disabled>Selecciona una serie
            </option>
            {
              series.map((serie) => {
                return (
                  <option key={serie.id} value={serie.id}>{serie.name}</option>
                )
              })
            }
          </select>
          { (errors.serie && showErrors) && (<p className='text-error'>{errors.serie}</p>) }
        </div>

        <div className='flex flex-col gap-2'>
          <label htmlFor="number" className='text-lg'>
              Escribe el número de volumen
          </label>
          <input
            min={0}
            type="number"
            name="number"
            id="number"
            className='input'
            placeholder='Número'
            value={ number }
            onChange={ onInputChange }
          />
          { (errors.number && showErrors) && (<p className='text-error'>{errors.number}</p>) }
        </div>
          
        <div className='flex flex-col gap-2'>
          <label htmlFor="price" className='text-lg'>
              Escribe el precio unitario
          </label>
          <input
            min={0}
            type="number"
            name="price"
            id="price"
            className='input'
            placeholder='Escribe el precio unitario'
            value={ price }
            onChange={ onInputChange }
          />
          { (errors.price && showErrors) && (<p className='text-error'>{errors.price}</p>) }
        </div>

        <div className='flex flex-col gap-2'>
          <label htmlFor="imgURL" className='text-lg'>
              Coloca la URL de la serie (cloudinary)
          </label>
          <input
            type="text"
            name="imgURL"
            id="imgURL"
            className='input'
            placeholder='Coloca una url de cloudinary'
            value={ imgURL }
            onChange={ onInputChange }
          />
          { (errors.imgURL && showErrors) && (<p className='text-error'>{errors.imgURL}</p>) }
        </div>

        <div className='flex flex-col gap-2'>
          <label htmlFor="title" className='text-lg'>
              Escribe un título
          </label>
          <input
            type="text"
            name="title"
            id="title"
            className='input'
            placeholder='Escribe un título'
            value={ title }
            onChange={ onInputChange }
          />
          { (errors.title && showErrors) && (<p className='text-error'>{errors.title}</p>) }
        </div> 

        <div className='grid grid-cols-2 gap-2'>
          <div className='flex flex-col'>
            <label htmlFor="month" className='text-lg'>
                Mes de publicación
            </label>
            <input
              min={1}
              max={12}
              type="number"
              name="month"
              id="month"
              className='input'
              value={ month }
              onChange={ onInputChange }
              placeholder='Introduce el mes de publicación'
            />
            { (errors.month && showErrors) && (<p className='text-error'>{errors.month}</p>) }
          </div>

          <div className='flex flex-col'>
            <label htmlFor="year" className='text-lg'>
                Año de publicación
            </label>
            <input
              min={2020}
              max={2030}
              type="number"
              name="year"
              id="year"
              className='input'
              value={ year }
              onChange={ onInputChange }
              placeholder='Introduce el año de publicación'
            />
            { (errors.year && showErrors) && (<p className='text-error'>{errors.year}</p>) }
          </div>
        </div>
        
      </div> 
      <button className='btn bg-accent w-full'>Guardar</button>
      <Toaster position='top-center'/>
    </form>
  )
}
