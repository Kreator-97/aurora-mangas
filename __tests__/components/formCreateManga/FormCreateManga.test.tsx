import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { series } from '../../fixtures/db'
import { FormCreateManga } from '../../../components/formCreateManga/FormCreateManga'

const onSubmit = jest.fn()

describe('Tests on FormCreateManga component', () => {
  const serie = series[0]

  test('should to match with snapshot', () => {
    const { container } = render(
      <FormCreateManga
        onSubmit={ async(formValues) => {}} 
        series={ series }
      />
    )
    
    expect(container).toMatchSnapshot()
  })

  test('should to contain all inputs necessary', () => {
    render(
      <FormCreateManga
        onSubmit={ async(formValues) => {}} 
        series={ series }
      />
    )
    const selectSerieLabel = screen.getByLabelText('Seleccionar serie')
    const numberLabel = screen.getByLabelText('Escribe el número de volumen')
    const unitPriceLabel = screen.getByLabelText('Escribe el precio unitario')
    const imgURLLabel = screen.getByLabelText('Coloca la URL de la serie (cloudinary)')
    const titleLabel = screen.getByLabelText('Escribe un título')
    const monthLabel = screen.getByLabelText('Mes de publicación')
    const yearLabel = screen.getByLabelText('Año de publicación')
    const stockLabel = screen.getByLabelText('Stock actual')
    const incrementStockLabel = screen.getByLabelText('Incrementar Stock')

    expect(selectSerieLabel).toBeInTheDocument
    expect(numberLabel).toBeInTheDocument
    expect(unitPriceLabel).toBeInTheDocument
    expect(imgURLLabel).toBeInTheDocument
    expect(titleLabel).toBeInTheDocument
    expect(monthLabel).toBeInTheDocument
    expect(yearLabel).toBeInTheDocument
    expect(stockLabel).toBeInTheDocument
    expect(incrementStockLabel).toBeInTheDocument
  })

  test('should to call onSubmit callback', async () => {
    render(
      <FormCreateManga
        onSubmit={ onSubmit }
        series={ series }
        resetOnSubmit={ false }
      />
    )

    const form = screen.getByTestId('form-create-manga')

    const selectSerie = screen.getByRole('combobox', {name: 'Seleccionar serie'})
    const numberInput = screen.getByRole('spinbutton', {name: 'Escribe el número de volumen'})
    const unitPriceInput = screen.getByRole('spinbutton', {name: 'Escribe el precio unitario'})
    const imgURLInput = screen.getByRole('textbox', {name: 'Coloca la URL de la serie (cloudinary)'})
    const title = screen.getByRole('textbox', {name: 'Escribe un título'})
    const monthInput = screen.getByRole('spinbutton', {name: 'Mes de publicación'})
    const yearInput = screen.getByRole('spinbutton', {name: 'Año de publicación'})
    const incrementStockLabel = screen.getByRole('spinbutton', {name: 'Incrementar Stock'})

    fireEvent.change(selectSerie, {target: { value: serie.id, name: 'serie' }})
    fireEvent.change(numberInput, {target: { value: '1', name: 'number' }})
    fireEvent.change(unitPriceInput, {target: { value: '99', name: 'price' }})
    fireEvent.change(imgURLInput, {target: { value: serie.imgURL, name: 'imgURL' }})
    fireEvent.change(title, {target: { value: 'volumen no.1', name: 'title' }})
    fireEvent.change(monthInput, {target: { value: '1', name: 'month' }})
    fireEvent.change(yearInput, {target: { value: '2022', name: 'year' }})
    fireEvent.change(incrementStockLabel, {target: { value: '100', name: 'incrementStock' }})
    
    fireEvent.submit(form, {preventDefault: () => {}})

    await waitFor(() => {

      expect(onSubmit).toHaveBeenCalledWith({
        imgURL: serie.imgURL,
        month: '1',
        number: '1',
        price: '99',
        serie: serie.id,
        title: 'volumen no.1',
        year: '2022',
        stock: 0,
        incrementStock: '100',
      })
    })
  })
})
