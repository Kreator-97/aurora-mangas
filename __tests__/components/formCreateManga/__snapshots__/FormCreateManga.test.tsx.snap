// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on FormCreateManga component should to match with snapshot 1`] = `
<div>
  <form
    class="mx-auto mb-4 max-w-lg"
    data-testid="form-create-manga"
  >
    <div
      class="flex flex-col gap-4 mb-4"
    >
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="text-lg"
          for="serie"
        >
          Seleccionar serie
        </label>
        <select
          class="input bg-white"
          id="serie"
          name="serie"
        >
          <option
            disabled=""
            value=""
          >
            Selecciona una serie
          </option>
          <option
            value="2baea5fb-2fd2-41ff-bf8b-f64d247a23dd"
          >
            Inuyasha
          </option>
          <option
            value="368852b4-62e7-449b-bd14-f6fd410037cc"
          >
            One Piece
          </option>
          <option
            value="543e5b82-6441-428e-b1c6-4c35db75d17b"
          >
            One Punch Man
          </option>
          <option
            value="9a4b8fa4-91cf-461f-a00c-efecfd3aae52"
          >
            Akira
          </option>
          <option
            value="9b63b0dc-adcd-40b9-8584-cdd3338555e1"
          >
            Hunter x Hunter
          </option>
        </select>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="text-lg"
          for="number"
        >
          Escribe el número de volumen
        </label>
        <input
          class="input"
          id="number"
          min="0"
          name="number"
          placeholder="Número"
          type="number"
          value=""
        />
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="text-lg"
          for="price"
        >
          Escribe el precio unitario
        </label>
        <input
          class="input"
          id="price"
          min="0"
          name="price"
          placeholder="Escribe el precio unitario"
          type="number"
          value="undefined"
        />
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="text-lg"
          for="imgURL"
        >
          Coloca la URL de la serie (cloudinary)
        </label>
        <input
          class="input"
          id="imgURL"
          name="imgURL"
          placeholder="Coloca una url de cloudinary"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="text-lg"
          for="title"
        >
          Escribe un título (opcional)
        </label>
        <input
          class="input"
          id="title"
          name="title"
          placeholder="Titulo opcional"
          type="text"
          value=""
        />
      </div>
      <div
        class="grid grid-cols-2 gap-2"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="text-lg"
            for="month"
          >
            Mes de publicación
          </label>
          <input
            class="input"
            id="month"
            max="12"
            min="1"
            name="month"
            placeholder="Introduce el mes de publicación"
            type="number"
            value=""
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-lg"
            for="year"
          >
            Año de publicación
          </label>
          <input
            class="input"
            id="year"
            max="2030"
            min="2020"
            name="year"
            placeholder="Introduce el año de publicación"
            type="number"
            value=""
          />
        </div>
      </div>
    </div>
    <button
      class="btn bg-accent w-full"
    >
      Guardar
    </button>
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
  </form>
</div>
`;
